lib/galaxy/model/deferred.py

```diff
@@ -0,0 +1,282 @@
+
+from galaxy.datatypes.sniff import (
+    convert_function,
+    stream_url_to_file,
+)
+from galaxy.exceptions import ObjectAttributeInvalidException
+from galaxy.files import ConfiguredFileSources
+from galaxy.model import (
+    Dataset,
+    DatasetCollection,
+    DatasetCollectionElement,
+    DatasetSource,
+    History,
+    HistoryDatasetAssociation,
+    HistoryDatasetCollectionAssociation,
+    LibraryDatasetDatasetAssociation,
+)
+from galaxy.objectstore import (
+    ObjectStore,
+    ObjectStorePopulator,
+)
+
+log = logging.getLogger(__name__)
+
+
+class TransientDatasetPaths(NamedTuple):
+    external_filename: str
+    external_extra_files_path: str
+
+
+class TransientPathMapper:
+    @abc.abstractmethod
+    def transient_paths_for(self, old_dataset: Dataset) -> TransientDatasetPaths:
+        """Map dataset to transient paths for detached HDAs.
+
+        Decide external_filename and external_extra_files_path that the supplied dataset's
+        materialized dataset should have its files written to.
+        """
+
+
+class SimpleTransientPathMapper(TransientPathMapper):
+    def __init__(self, staging_directory: str):
+        self._staging_directory = staging_directory
+
+    def transient_paths_for(self, old_dataset: Dataset) -> TransientDatasetPaths:
+        external_filename_basename = "dataset_%s.dat" % str(old_dataset.uuid)
+        external_filename = os.path.join(self._staging_directory, external_filename_basename)
+        external_extras_basename = "dataset_%s_files" % str(old_dataset.uuid)
+        external_extras = os.path.join(self._staging_directory, external_extras_basename)
+        return TransientDatasetPaths(external_filename, external_extras)
+
+
+class DatasetInstanceMaterializer:
+    """This class is responsible for ensuring dataset instances are not deferred."""
+
+    def __init__(
+        self,
+        attached: bool,
+        object_store_populator: Optional[ObjectStorePopulator] = None,
+        transient_path_mapper: Optional[TransientPathMapper] = None,
+        file_sources: Optional[ConfiguredFileSources] = None,
+        sa_session: Optional[scoped_session] = None,
+    ):
+        """Constructor for DatasetInstanceMaterializer.
+
+        If attached is true, these objects should be created in a supplied object store.
+        If not, this class produces transient HDAs with external_filename and
+        external_extra_files_path set.
+        """
+        self._attached = attached
+        self._transient_path_mapper = transient_path_mapper
+        self._object_store_populator = object_store_populator
+        self._file_sources = file_sources
+        self._sa_session = sa_session
+
+    def ensure_materialized(
+        self,
+        dataset_instance: Union[HistoryDatasetAssociation, LibraryDatasetDatasetAssociation],
+        target_history: Optional[History] = None,
+    ) -> HistoryDatasetAssociation:
+        """Create a new detached dataset instance from the supplied instance.
```

Comment:
It is both - the detached instances are not written to an object store or the database. A "detached" (in the SQL Alchemy sense) HDA is created for tool evaluation that has a direct path encoded instead of relying on an object store. When we do the remote evaluation thing we want it to be detached in both senses right?